### PR TITLE
feat: add scheduled delivieres interactive users

### PR DESCRIPTION
**Pull request by rephus**
Closes: #8551

### Description:

Charts: 
Screenshot from 2024-02-01 10-46-38
Dashboards:

Screenshot from 2024-02-01 10-57-59

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

**Comment by rephus** [comment]
Syncs are out of scope

File: packages/backend/src/services/DashboardService/DashboardService.ts [hunk 1]
```diff
@@ -665,8 +665,11 @@ export class DashboardService {
         const { organizationUuid, projectUuid } = dashboard;
         if (
             user.ability.cannot(
-                'update',
-                subject('Dashboard', { organizationUuid, projectUuid }),
+                'manage',
+                subject('ScheduledDeliveries', {
```
Comment (ZeRego) [inline packages/backend/src/services/DashboardService/DashboardService.ts:669]: We should rename the method to be more explicit as it was initially intended to check dashboard access.

File: packages/backend/src/services/SavedChartsService/SavedChartService.ts [hunk 2]
```diff
@@ -80,8 +80,11 @@ export class SavedChartService {
         const { organizationUuid, projectUuid } = savedChart;
         if (
             user.ability.cannot(
-                'update',
-                subject('SavedChart', { organizationUuid, projectUuid }),
+                'manage',
+                subject('ScheduledDeliveries', {
```
Comment (ZeRego) [inline packages/backend/src/services/SavedChartsService/SavedChartService.ts:84]: Note that this method was intended to check chart access and is used in methods not related to scheduled delivery.

File: examples/full-jaffle-shop-demo/dbt/models/events.yml [hunk 3]
```diff
@@ -48,15 +48,7 @@ models:
             urls:
               - label: Open in Google Calendar
                 url: 'https://calendar.google.com/calendar/u/0/r/day/${ value.formatted |split: "-" |join: "/"}'
-      - name: event
```
Comment (ZeRego) [inline examples/full-jaffle-shop-demo/dbt/models/events.yml:51]: where these changes intentional?

Comment (rephus) [reply examples/full-jaffle-shop-demo/dbt/models/events.yml:51]: nope, sorry, rolling back

File: packages/backend/src/services/SavedChartsService/SavedChartService.ts (hunk 2 above)
Comment (rephus) [reply packages/backend/src/services/SavedChartsService/SavedChartService.ts:84]: ok I can create a separate check for scheduled deliveries then

File: packages/common/src/authorization/projectMemberAbility.ts [hunk 4]
```diff
@@ -42,6 +42,9 @@ export const projectMemberAbilities: Record<
         can('manage', 'ChangeCsvResults', {
             projectUuid: member.projectUuid,
         });
+        can('manage', 'ScheduledDeliveries', {
+            projectUuid: member.projectUuid,
```
Comment (ZeRego) [inline packages/common/src/authorization/projectMemberAbility.ts:46]: So now interactive viewers can edit any scheduled delivery, even the ones created by editors? doesn't feel right 🤔 @TuringLovesDeathMetal

Comment (hamzahc1) [reply packages/common/src/authorization/projectMemberAbility.ts:46]: I couldn't find a linked Slack thread for this issue in Cloudy - where did this request come from? It'd be useful context

I've heard a version of this that viewers should be able to create their own deliveries - but editing ones that an admin has created doesn't seem right 🤔

Comment (TuringLovesDeathMetal) [reply packages/common/src/authorization/projectMemberAbility.ts:46]: Yes, sorry, this was a request from Workday. I asked them about the permission to edit any scheduled delivery, and they said "yes", so I just added that to the permissions. 

Here's the very small info about this specific thing from our call: https://www.notion.so/lightdash/Workday-20dcba7760ca4ea09f999d24a6022330?pvs=4#d60f2b2653264220a8c5f1928c4f23fc (it's also in their shared feature doc). 

I initially also thought "no editing existing deliveries", but then the UX became more complicated and Workday said they didn't want that, so I updated the ticket.

Comment (rephus) [reply packages/common/src/authorization/projectMemberAbility.ts:46]: I updated this PR, you can view admin schedulers, but you can't update/delete them. This error is happening in the backend for now.

File: packages/backend/src/services/SavedChartsService/SavedChartService.ts (hunk 2 above)
Comment (rephus) [reply packages/backend/src/services/SavedChartsService/SavedChartService.ts:109]: done

File: packages/backend/src/services/DashboardService/DashboardService.ts (hunk 1 above)
Comment (rephus) [reply packages/backend/src/services/DashboardService/DashboardService.ts:667]: done

File: packages/backend/src/services/SchedulerService/SchedulerService.ts [hunk 5]
```diff
@@ -83,26 +83,35 @@ export class SchedulerService {
...
+            }),
+        );
+        const isDeliveryOwner = scheduler.createdBy === user.userUuid;
+
+        if (canManageDeliveries || (canCreateDeliveries && isDeliveryOwner)) {
+            return { scheduler, resource };
         }
-        return { scheduler, resource };
+
+        throw new ForbiddenError();
```
Comment (ZeRego) [inline packages/backend/src/services/SchedulerService/SchedulerService.ts:114]: We should move all this logic to the permission. 

viewer
```
        can('create', 'ScheduledDeliveries', {
            projectUuid: member.projectUuid,
        });
        can('manage', 'ScheduledDeliveries', {
            projectUuid: member.projectUuid,
            createdBy: member.userUuid,
        });
```

editor (nothing changes)
```
        can('manage', 'ScheduledDeliveries', {
            projectUuid: member.projectUuid,
        });
```

The service and fe can now check permissions with 
```
       const canCreateScheduleDelivery = user.ability.can(
            'create',
            subject('ScheduledDeliveries', project),
        );

        const canManageScheduleDelivery = user.ability.can(
            'manage',
            subject('ScheduledDeliveries', scheduler),
        );
```

This is similar to what we specced for space permissions.

**Review by ZeRego — APPROVED** [review]

File: packages/backend/src/services/SchedulerService/SchedulerService.ts (hunk 5 above)
Comment (ZeRego) [reply packages/backend/src/services/SchedulerService/SchedulerService.ts:114]: Can be done in separate PR with the improvements to the FE.

**Comment by rephus** [comment]
api error caused by a bad initial migration , tested locally

image

**Pull request merged**